### PR TITLE
Support backticks in method names

### LINKED ISSUE #1621: Buck blows up on backticks in kotlin
If you have a backtick in an API for kotlin, buck will fail with a stacktrace like this.

Example: `fun 'in'()` (replace the `'` with backticks, github won't format it nicely)

```
./buckw test //libraries/foundation/realtime/realtime-projectgen/... --include unit                                                                                    
Buck encountered an internal error
java.lang.IllegalStateException
	at com.google.common.base.Preconditions.checkState(Preconditions.java:429)
	at com.facebook.buck.jvm.java.AccumulateClassNamesStep.parseClassHashes(AccumulateClassNamesStep.java:167)
	at com.facebook.buck.jvm.java.JavaLibraryRules.initializeFromDisk(JavaLibraryRules.java:81)
	at com.facebook.buck.jvm.java.DefaultJavaLibrary.initializeFromDisk(DefaultJavaLibrary.java:333)
	at com.facebook.buck.jvm.java.DefaultJavaLibrary.initializeFromDisk(DefaultJavaLibrary.java:1)
	at com.facebook.buck.rules.BuildOutputInitializer.initializeFromDisk(BuildOutputInitializer.java:47)
	at com.facebook.buck.rules.CachingBuildRuleBuilder.doInitializeFromDisk(CachingBuildRuleBuilder.java:1276)
	at com.facebook.buck.rules.CachingBuildRuleBuilder.finalizeBuildRule(CachingBuildRuleBuilder.java:403)
	at com.facebook.buck.rules.CachingBuildRuleBuilder.lambda$4(CachingBuildRuleBuilder.java:288)
	at com.facebook.buck.rules.CachingBuildRuleBuilder.lambda$21(CachingBuildRuleBuilder.java:1634)
	at com.google.common.util.concurrent.AbstractTransformFuture$AsyncTransformFuture.doTransform(AbstractTransformFuture.java:211)
	at com.google.common.util.concurrent.AbstractTransformFuture$AsyncTransformFuture.doTransform(AbstractTransformFuture.java:200)
	at com.google.common.util.concurrent.AbstractTransformFuture.run(AbstractTransformFuture.java:130)
	at com.google.common.util.concurrent.MoreExecutors$5$1.run(MoreExecutors.java:952)
	at com.facebook.buck.util.concurrent.WeightedListeningExecutorService.lambda$2(WeightedListeningExecutorService.java:104)
	at com.facebook.buck.util.concurrent.WeightedListeningExecutorService.lambda$0(WeightedListeningExecutorService.java:78)
	at com.google.common.util.concurrent.AbstractTransformFuture$AsyncTransformFuture.doTransform(AbstractTransformFuture.java:211)
	at com.google.common.util.concurrent.AbstractTransformFuture$AsyncTransformFuture.doTransform(AbstractTransformFuture.java:200)
	at com.google.common.util.concurrent.AbstractTransformFuture.run(AbstractTransformFuture.java:130)
	at com.google.common.util.concurrent.MoreExecutors$5$1.run(MoreExecutors.java:952)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:748)

    When building rule //libraries/foundation/realtime/realtime-projectgen:test_main#testsjar.
Downloaded 68.87 Kbytes/sec avg, 3 artifacts, 48.00 Kbytes
Building: finished in 2.1 sec (100%) 397/398 jobs, 1 updated, 0.3% cache miss
  Total time: 2.3 sec
```

This is kind of serious, as backticks in kotlin are quite normal to avoid keyword collisions

### PR DESCRIPTION
**Pull request by NickFirmani**
Resolves #1621

When a method contains backticks (in Kotlin) the resultant generated `classes.txt` file has lines that look like the following: 
```
com/Example/ExampleTest$example method name$1 e925a3f
```

Since spaces are valid in this case, splitting into two parts and aborting if there are more than two parts isn't the ideal behavior, so instead we split on the last space in the line.

### CODE REVIEW

File: src/com/facebook/buck/jvm/java/AccumulateClassNamesStep.java [hunk 1]
```diff
@@ -163,10 +158,9 @@ public InputStream openStream() throws IOException {
     Map<String, HashCode> classNames = new HashMap<>();
 
     for (String line : lines) {
-      List<String> parts = CLASS_NAME_AND_HASH_SPLITTER.splitToList(line);
-      Preconditions.checkState(parts.size() == 2);
-      String key = parts.get(0);
-      HashCode value = HashCode.fromString(parts.get(1));
+      int index = line.trim().lastIndexOf(CLASS_NAME_HASH_CODE_SEPARATOR);
```
Comment (philipjameson) [inline src/com/facebook/buck/jvm/java/AccumulateClassNamesStep.java]: The trim probably isn't necessary and creates a copy. Just precondition check on index > 0

File: src/com/facebook/buck/jvm/java/AccumulateClassNamesStep.java [hunk 2]
```diff
@@ -163,10 +159,11 @@ public InputStream openStream() throws IOException {
     Map<String, HashCode> classNames = new HashMap<>();
 
     for (String line : lines) {
-      List<String> parts = CLASS_NAME_AND_HASH_SPLITTER.splitToList(line);
-      Preconditions.checkState(parts.size() == 2);
-      String key = parts.get(0);
-      HashCode value = HashCode.fromString(parts.get(1));
+      int index = line.lastIndexOf(CLASS_NAME_HASH_CODE_SEPARATOR);
+      Preconditions.checkArgument(
+          index > 0, String.format("Class names and hashcodes malformed: %1$s", line));
```
Comment (philipjameson) [inline src/com/facebook/buck/jvm/java/AccumulateClassNamesStep.java]: Can you please use the form of checkArgument that takes a format string + arguments? We don't want to allocate a string each time that we are in this loop for the ideally non-occurring case that index < 0

https://guava.dev/releases/21.0/api/docs/com/google/common/base/Preconditions.html#checkArgument-boolean-java.lang.String-java.lang.Object...-

File: test/com/facebook/buck/jvm/kotlin/KotlinLibraryIntegrationTest.java [hunk 3]
```diff
@@ -161,6 +161,12 @@ public void shouldFailToCompileInvalidKotlinCode() {
     buildResult.assertFailure();
   }
 
+  @Test
+  public void shouldCompileKotlinCodeWithBackticks() {
+    ProcessResult buildResult = workspace.runBuckCommand("build", "//com/example/backticks:example");
```
Comment (philipjameson) [inline test/com/facebook/buck/jvm/kotlin/KotlinLibraryIntegrationTest.java]: The general pattern here is to just do `workspace.runBuckBuild("//com/example/backticks:example").assertSuccess();`

**Comment by NickFirmani** [comment]
Thanks for the reviews!

**Comment by philipjameson** [comment]
Thanks. I’ll take a look and get this imported later this evening then

**Comment by philipjameson** [comment]
Hey @NickFirmani, I patched in your diff, and it seems like your test is passing without your changes? Can you please verify?

**Comment by NickFirmani** [comment]
Hey @philipjameson, I have been poking around at that a bit, the problem seems to be that the `AccumulateClassNamesStep` doesn't run on the integration test with the way it is setup. I'm not super familiar with this, but will keep looking when I next have time

**Comment by v-jizhang** [comment]
BUCK supports backticks in method names already.

**Pull request closed without merging**